src/mapport.cpp

```diff
@@ -0,0 +1,236 @@
+static void ThreadUpnp()
+{
+    std::string port = strprintf("%u", GetListenPort());
+    const char * multicastif = nullptr;
+    const char * minissdpdpath = nullptr;
+    struct UPNPDev * devlist = nullptr;
+    char lanaddr[64];
+
+    int error = 0;
+#if MINIUPNPC_API_VERSION < 14
+    devlist = upnpDiscover(2000, multicastif, minissdpdpath, 0, 0, &error);
+#else
+    devlist = upnpDiscover(2000, multicastif, minissdpdpath, 0, 0, 2, &error);
+#endif
+
+    struct UPNPUrls urls;
+    struct IGDdatas data;
+    int r;
+
+    r = UPNP_GetValidIGD(devlist, &urls, &data, lanaddr, sizeof(lanaddr));
+    if (r == 1)
+    {
+        if (fDiscover) {
+            char externalIPAddress[40];
+            r = UPNP_GetExternalIPAddress(urls.controlURL, data.first.servicetype, externalIPAddress);
+            if (r != UPNPCOMMAND_SUCCESS) {
+                LogPrintf("UPnP: GetExternalIPAddress() returned %d\n", r);
+            } else {
+                if (externalIPAddress[0]) {
+                    CNetAddr resolved;
+                    if (LookupHost(externalIPAddress, resolved, false)) {
+                        LogPrintf("UPnP: ExternalIPAddress = %s\n", resolved.ToString());
+                        AddLocal(resolved, LOCAL_MAPPED);
+                    }
+                } else {
+                    LogPrintf("UPnP: GetExternalIPAddress failed.\n");
+                }
+            }
+        }
+
+        std::string strDesc = PACKAGE_NAME " " + FormatFullVersion();
+
+        do {
+            r = UPNP_AddPortMapping(urls.controlURL, data.first.servicetype, port.c_str(), port.c_str(), lanaddr, strDesc.c_str(), "TCP", 0, "0");
+
+            if (r != UPNPCOMMAND_SUCCESS) {
+                LogPrintf("AddPortMapping(%s, %s, %s) failed with code %d (%s)\n", port, port, lanaddr, r, strupnperror(r));
+            } else {
+                LogPrintf("UPnP Port Mapping successful.\n");
+            }
+        } while (g_mapport_interrupt.sleep_for(std::chrono::minutes(20)));
+
+        r = UPNP_DeletePortMapping(urls.controlURL, data.first.servicetype, port.c_str(), "TCP", 0);
+        LogPrintf("UPNP_DeletePortMapping() returned: %d\n", r);
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        FreeUPNPUrls(&urls);
+    } else {
+        LogPrintf("No valid UPnP IGDs found\n");
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        if (r != 0)
+            FreeUPNPUrls(&urls);
+    }
+}
+#endif // #ifdef USE_UPNP
+
+void StartMapPort(MapPort proto)
+{
+    if (!g_mapport_thread.joinable()) {
+        assert(!g_mapport_interrupt);
+        switch (proto) {
+        case MapPort::NAT_PMP: {
+#ifdef USE_NATPMP
+            g_mapport_thread = std::thread((std::bind(&TraceThread<void (*)()>, "natpmp", &ThreadNatpmp)));
+#endif // #ifdef USE_NATPMP
+            return;
+        }
+        case MapPort::UPNP: {
+#ifdef USE_UPNP
+            g_mapport_thread = std::thread((std::bind(&TraceThread<void (*)()>, "upnp", &ThreadUpnp)));
```

Comment:
Won't this overwrite the `g_mapport_thread` set above if both are enabled?

Reply:
I believe that the both port mapping protocols shouldn't be enabled simultaneously:
https://github.com/bitcoin/bitcoin/blob/c7e90ca40ad9b7c66f3429b792435e9279ceca6b/src/init.cpp#L1781-L1786

Reply:
It should be possible to enable both.

Reply:
If both are enabled, would you try to open de port with NAT_PMP first and if that fails UPNP?

Reply:
> It should be possible to enable both.

Do you mean that a router could do port mapping using both NAT-PMP and UPnP simultaneously? What is reason for that?

Reply:
I wouldn't use them simultaneously, but one fails it makes sense to try the other. Though I believe our long plan is to remove UPnP, if NAT-PMP does the job.

Reply:
> I wouldn't use them simultaneously, but one fails it makes sense to try the other.

Good suggestion for a followup.

> Though I believe our long plan is to remove UPnP, if NAT-PMP does the job.

Not all routers with stock software have NAT-PMP port mapping feature.

Reply:
Users who don't know what their router supports should be able to just flip both on. And laptops might very well migrate between different routers using each protocol.

Agree with @hebasto that NAT-PMP can't be a complete substitute for UPnP since routers might not support both.

Reply:
Addressed in the latest push.